### PR TITLE
Update SmallRye OpenAPI and MicroProfile OpenAPI to 4.0

### PR DESCRIPTION
**Pull request by MikeEdgar**
* Bumps MicroProfile OpenAPI to 4.0.2 (breaking API changes)
* Bumps SmallRye OpenAPI to 4.0.0
* Replace uses of SmallRye OpenAPI's internal model types with `OASFactory`
* Move calculation of operation method reference into OpenAPI processor and combine two Operation filters into single filter to make a single pass over the OpenAPI model.

### CODE REVIEW

File: bom/application/pom.xml [hunk 1]
```diff
@@ -46,12 +46,12 @@
...
         <microprofile-jwt.version>2.1</microprofile-jwt.version>
         <microprofile-lra.version>2.0</microprofile-lra.version>
-        <microprofile-openapi.version>3.1.2</microprofile-openapi.version>
+        <microprofile-openapi.version>4.0.2</microprofile-openapi.version>
         <smallrye-common.version>2.8.0</smallrye-common.version>
         <smallrye-config.version>3.10.0</smallrye-config.version>
         <smallrye-health.version>4.1.0</smallrye-health.version>
         <smallrye-metrics.version>4.0.0</smallrye-metrics.version>
-        <smallrye-open-api.version>3.13.0</smallrye-open-api.version>
+        <smallrye-open-api.version>4.0.0-beta2</smallrye-open-api.version>
```
Comment (zakkak) [inline bom/application/pom.xml:54]: This is just for testing purposes right? (and the reason the PR is a draft I guess)

Comment (MikeEdgar) [reply bom/application/pom.xml:54]: Yes, exactly. The 4.0.0 release should be out in a a few days but it won't really change much about this PR other than this version property.

**Comment by zakkak** [comment]
For reference, this PR is related to https://github.com/quarkusio/quarkus/pull/43616 and https://github.com/quarkusio/quarkus/pull/41037.

The mentioned reduction in the native image size refers to https://github.com/quarkusio/quarkus/pull/41037#issuecomment-2348117351 and how https://github.com/quarkusio/quarkus/pull/43616 reverted some of that effect.

**Comment by MikeEdgar** [comment]
I'm looking into the integration test failure.

**Comment by MikeEdgar** [comment]
@phillip-kruger please take a look when you can, thanks.

**Review by phillip-kruger — APPROVED** [review]
Nice ! Thanks @MikeEdgar !

**Comment by zakkak** [comment]
> @zakkak, you should see further reduction in the native image size with this update.

Indeed, I see:

| | Before | After | Diff (%) |
| --- | --- | --- | --- |
| Reachable classes | 13680 | 13421 | 1.9 |
| Reachable methods | 70690 | 69608 | 1.5 |
| **Size (MB)** | **63.6** | **63.1** | **0.7** |

**Review by gsmet — APPROVED** [review]

**Comment by michalvavrik** [comment]
Hello, this PR has in description `breaking changes` but there is no label and there isn't a migration guide note. For example, JSON path to one of my example in schema was `components.schemas.LocalDateTime.example` and now it is `components.schemas.LocalDateTime.example. This will impact generated clients, ETLs etc. and users need to know about breaking changes. @MikeEdgar can you document this, please? If you cannot add migration guides, I am sure @gsmet or @phillip-kruger can help.

**Comment by MikeEdgar** [comment]
> @MikeEdgar can you document this, please?

@michalvavrik can you point me to the repository where that should be done?

**Comment by michalvavrik** [comment]
> > @MikeEdgar can you document this, please?
> 
> @michalvavrik can you point me to the repository where that should be done?

It's this one. I myself have no rights so better talk to @gsmet about the process.

**Comment by michalvavrik** [comment]
I'll add a breaking label though. Hope that's fine.

**Comment by zakkak** [comment]
> @michalvavrik can you point me to the repository where that should be done?

@MikeEdgar the migration guide is here https://github.com/quarkusio/quarkus/wiki/Migration-Guide-3.17 , I don't know if you have edit rights through.

**Comment by gsmet** [comment]
You can post something here and I'll add it.

**Comment by MikeEdgar** [comment]
@gsmet , @michalvavrik , here is some content for the 3.17 Migration guide.

## MicroProfile OpenAPI 4.0

MicroProfile OpenAPI and the SmallRye OpenAPI implementation have both be updated to version 4. This version brings support for [OpenAPI specification verion 3.1](https://spec.openapis.org/oas/v3.1.0.html). By default, the OpenAPI documentation generated will use version `3.1.0`, however users can continue to generate `3.0.x` output by using property `mp.openapi.extensions.smallrye.openapi` with (for example) the value `3.0.3`.

The `Schema` interface contains breaking changes, specifically:
- `type` property type changed from `SchemaType` to `List<SchemaType>`
- `exclusiveMinimum` and `exclusiveMaximum` property types changed from `Boolean` to `BigDecimal`
- `nullable` property removed (replaced by the addition of `NULL` to `SchemaType`)
- property `example` is deprecated in favor of using the new `examples` property, which may hold one or more example objects

Additionally, the default value of `@RequestBody.required` has changed to `true` to reflect that this is the much more common case where a RESTful resource method accepts a request body.

There are numerous other [new properties for annotations and models](https://download.eclipse.org/microprofile/microprofile-open-api-4.0.2/microprofile-openapi-spec-4.0.2.html#api_changes_40), reflecting the changes between OpenAPI 3.0.x and 3.1.0. The release notes for the latest MicroProfile OpenAPI 4 release are [available](https://download.eclipse.org/microprofile/microprofile-open-api-4.0.2/microprofile-openapi-spec-4.0.2.html#release_notes_40) for review.

**Comment by MikeEdgar** [comment]
@gsmet just bumping this thread to make sure you got my comment with the 3.17 migration notes for OpenAPI.

**Pull request merged**